### PR TITLE
DEVEN-2894 Invoke shellHook

### PR DESCRIPTION
**Pull request by yanw-canva**
## Context

cached-nix-shell doesn't currently invoke shellHook after the first run.

## Intent
Invoke shellHook on interactive mode and shebang mode.

### CODE REVIEW

File: src/main.rs [hunk 1]
```diff
@@ -280,9 +283,21 @@ fn run_script(
...
     let exec = if is_literal_bash_string(nix_shell_args.interpreter.as_bytes())
     {
-        Command::new(nix_shell_args.interpreter)
-            .arg(fname)
-            .args(script_args)
+        // eprintln!("Interpreter is a literal string, executing directly");
+        // We also check for CNS_CACHED_UPDATED so that we don't run shellHook twice when
+        // the cache is populated/updated after running real nix-shell (which will invoke shellHook)
+        let mut exec_string = OsString::new();
+        exec_string.push(r#"[ -z "${CNS_CACHED_UPDATED}" -a -n "${shellHook:-}" ] && eval "$shellHook"; exec "#);
```
Comment (olebedev) [inline src/main.rs]:
```suggestion
        exec_string.push(r#"[ -z "${CNS_SHELL_HOOKED}" -a -n "${shellHook:-}" ] && eval "$shellHook"; exec "#);
```

File: src/main.rs [hunk 2]
```diff
@@ -411,7 +426,11 @@ fn cached_shell_env(pure: bool, inp: &NixShellInput) -> EnvOptions {
         cache_write(&inputs_hash, "trace", &outp.trace.serialize());
         cache_symlink(&inputs_hash, "drv", &outp.drv);
 
-        outp.env
+        // set an environment variable to mark that cache is populated/updated
+        let mut cached_env = outp.env;
+        cached_env.insert(OsString::from("CNS_CACHED_UPDATED"), OsString::from("1"));
```
Comment (olebedev) [inline src/main.rs]:
```suggestion
        cached_env.insert(OsString::from("CNS_SHELL_HOOKED"), OsString::from("1"));
```

**Review by olebedev — APPROVED** [review]
LGTM with the suggestions, please have a look below.

Not blocking: maybe considering using a template engine for our bash scripting. Any thoughts or suggestions on this?

**Comment by yanw-canva** [comment]
For sure template engine would make composing bash commands cleaner. Tera or Handlebars seem like they can do the job, but I'm new to Rust so I don't have opinion on which one to use. But this project is small enough I think we can switch template engines without much difficulties if we have to.

**Pull request merged**